Split classic boot configuration

waveform80 opened the pull request:
This is a fairly complex PR that moves the classic branch closer to the core branch in preparation for eventual merging (although this is not yet that merge as other things need to happen in snapd and flash-kernel for that to occur), and splits the boot configuration for easier maintenance. Specifically, this PR:

* Splits the boot configuration into a series of configuration files, most of which are not (intended to be) user-editable
* Instead of copying a single u-boot binary to the boot partition, copies all u-boot binaries from the u-boot package under different filenames (which the new config.txt will select according to platform); this is how core already does things (albeit from a different source, as core builds its own u-boot binaries)
* Uses the boot.scr uboot script from the flash-kernel package instead of a (redundant) one that was included in this branch (this is another move in preparation for merging with core as the pi boot script in flash-kernel will eventually support both platforms)
* Includes the option to use a different series than the builder's for packages

For review, I would recommend going through the commits individually then looking at the end state rather than trying to review the overall diff.

waveform80 committed:
Split the boot configuration

The main config.txt needs to call different u-boot binaries for
different pi revisions (mostly to support the pi4). In addition, it's
desirable to have config.txt essentially static with modifiable
includes. This commit moves to this model and also includes some extra
configuration examples.

In concert with this change, the u-boot-rpi package will be updated to
include a postinst which will migrate existing installations to this
style of configuration.

After testing it appears that on older firmwares conditional [sections]
aren't recognized. Hence the bootloader will read the [pi3] section and
then carry on to read the [pi4] section! Naturally, the pi4 u-boot isn't
optimized for use on the pi3, but more importantly the pi4 u-boot won't
exist at all on images built with older u-boot-rpi packages.

As a workaround, specify kernels in an unusual order (4 first, then 2,
then 3) so that later settings override earlier ones. This results in
the correct kernel (uboot) selection with both old and new firmwares
across platforms.

waveform80 committed:
Copy multiple u-boots per image

waveform80 committed:
Use boot script from flash-kernel

Supports both archs and compression on each (which the bootscript in
this repo doesn't).

NOTE: The RPi bootscript was named bootscr.rpi3 in bionic (along with a
ubootenv for the pi2); in subsequent releases this simply became a
generic bootscr.rpi. This explains the "*" wildcard on the source of
mkimage.

waveform80 committed:
Tidying up

No need to copy the boot.scr all over the place - just stuff it in
boot-assets like everything else. Also make all copies to the boot
partition consistently use "-a" (probably pointlessly given it's FAT,
but still...)

waveform80 committed:
Use the specified series instead of the host's

waveform80 committed:
Strange issue with serial when BT is disabled

Serial console works with BT disabled, but only for rx; tx seems
disabled?! Needs further investigation...

sil2100 commented on Makefile:
```diff
@@ -51,43 +49,42 @@ all: clean
 	# Preparation stage
 	mkdir -p $(STAGEDIR)/debs $(STAGEDIR)/unpack
-	# u-boot
+	$(call stage_package,flash-kernel,$(STAGEDIR),$(ARCH))
```
Good to see we'll be finally pulling the bootscript from a common place (flash-kernel). Just a note here: before we can actually merge this, we'll probably need to SRU flash-kernel to bionic, since currently it doesn't have the bootscr.rpi in it. Even though the next bionic point release is still in some time, we do build dailies for that series. So I'd say we'd want flash-kernel in at least bionic-proposed.

waveform80 replied:
Actually I don't think this is an issue. There's a very subtle change in the script which deals with this different. Specifically, Eoan (and Disco) have bootscript/all/bootscr.rpi in flash-kernel, while Bionic has bootscript/all/bootscr.rpi3 (and a uboot.env for pi2) in its flash-kernel. Now, on line 66 note that mkimage is called with `.../bootscr.rpi*` rather than `.../bootscr.rpi` to account for this difference (yes, it's a bit hacky to use a wildcard here but I think it's acceptable given the requirement).

Incidentally, this is (one of the reasons) why I suggested going through the individual commits rather than just looking at the end state; 17596c4f3458b4e368224453f1defe0ce2bfaa69 mentions this case in the commit message.

sil2100 replied:
hm, you are right and flash-kernel from bionic does have the base bootscript for the pi3. But I still think there will be a problem right now - a bit of a different one though. So the builds will not FTBFS, but I suppose our arm64 bionic daily images will stop working.

The old flash-kernel bootscr.rpi3 that we have there I guess is very armhf oriented (which is why it's in 'armhf' instead of 'all'), supporting only booting the armhf kernel - since there are no conditional checks there for now. Certainly this is less of a blocker, as we never really give any guarantee of bootability of our daily images, but it's something we should sort out earlier before the point release. So I'd say let's aim for getting it into bionic-proposed at least in the nearest time.

sil2100 commented on configs/syscfg.txt:
```diff
@@ -0,0 +1,9 @@
+dtparam=i2c_arm=on
+dtparam=spi=on
+
+include nobtcfg.txt
```
Ah, so this is the place which the bluetooth package will modify to switch bluetooth on and off, right?

waveform80 replied:
This is indeed the intention; flip this to `btcfg.txt` and back to switch between the configs (and eventually we'll have a CLI tool to handle this and the manipulation of the dtparam lines above).

sil2100 commented on configs/README:
```diff
@@ -0,0 +1,26 @@
+An overview of the files on the /boot/firmware partition (the 1st partition
+on the SD card) used by the Ubuntu boot process (roughly in order) is as
+follows:
+
+* bootcode.bin   - this is the second stage bootloader loaded by all pis with
+                   the exception of the pi4 (where this is replaced by flash
+                   memory)
+* config.txt     - the first configuration file read by the boot process
+* syscfg.txt     - the file in which system modified configuration will be
+                   placed, included by config.txt
+* usercfg.txt    - the file in which user modified configuration should be
+                   placed, included by config.txt
+* start*.elf     - the third stage bootloader, which handles device-tree
+                   modification and which loads...
+* uboot*.bin     - various u-boot binaries for different pi platforms; these
+                   are launched as the "kernel" by config.txt
+* boot.scr       - the boot script executed by uboot*.bin which in turn
+                   loads...
+* vmlinuz        - the Linux kernel, executed by boot.scr
+* initrd.img     - the initramfs, executed by boot.scr
+* meta-data      - meta-data for cloud-init; usually just contains the
+                   instance id
+* network-config - network configuration for cloud-init; edit this to set up
+                   wifi access points and other networking settings
+* user-data      - user-data for cloud-init; edit this to configure initial
+                   users, SSH keys, packages, etc.
```
Love this documentation - this is something we've been missing even before. Thanks for adding it!

sil2100 reviewed (approved):
+1 on the changes. I will not merge it yet as per the inline comment regarding flash-kernel. But other than that, this looks like good to go (as I'm sure you tested this on a live system).

waveform80 committed:
The new Pi4 kernel requires the v3d overlay

Otherwise the framebuffer doesn't initialize correctly

mvo5 commented:
A bit of a meta-question: once we are happy with things here we will port to "pi2-gadget", "cm3-gadget" and "pi-gadget" (the  last one is the unified pi gadget we use in core18) ?

waveform80 commented:
> A bit of a meta-question: once we are happy with things here we will port to "pi2-gadget", "cm3-gadget" and "pi-gadget" (the last one is the unified pi gadget we use in core18) ?

Well, this PR is only a step towards unification of the classic and core gadgets; there's a bit more work to be done (including within snapd itself) before that can be achieved. If pi-gadget is the unified one it may be worth continuing the effort there (on the core side of things); whether it's worth merging further work into pi2-gadget and cm3-gadget I don't know. My understanding was that the eventual plan was to retire those in favour of a unified image?

waveform80 committed:
Use vc4-fkms-v3d on all platforms

Framebuffer is corrupt otherwise (invalid mode setting by the looks of
it). Ideally we want to run without this overlay but it appears the
current pi4-compatible kernel requires this on all platforms (and not,
as previously thought, just the Pi4).
